Don't log connection closed exceptions as ERROR in websockets-next

geoand opened the pull request:
The clients can choose to close at any time, so it does not make sense to fill up logs with closed connection messages

mkouba reviewed (commented):
> The clients can choose to close at any time, so it does not make sense to fill up logs with closed connection messages

I can't say that I agree with this proposal.

When a client closed a connection then the callback should not be called at all. If it is then it should only happen infrequently (when the connection is closed during processing or something like that). When exactly do you observe the logs filled with "closed connection" messages?

Also a user can define a global error handler to filter out these exceptions.

geoand commented on integration-tests/opentelemetry-scheduler/src/main/java/io/quarkus/it/opentelemetry/scheduler/Counter.java:
```diff
@@ -23,6 +23,7 @@ public int get() {
 
     @Scheduled(cron = "*/1 * * * * ?", identity = "myCounter")
     void increment() throws InterruptedException {
+        System.out.println(Thread.currentThread().getName());
```
Left over from a previous section :)

geoand commented on extensions/websockets-next/server/runtime/src/main/java/io/quarkus/websockets/next/runtime/WebSocketServerRecorder.java:
```diff
@@ -254,6 +253,29 @@ public void handle(Void event) {
+        if (message == null) {
+            return false;
+        }
+        return message.contains("WebSocket is closed");
```
The exception is a `NoStackTraceException` so there is nothing to go by.

I do agree it's brittle, but I don't see any alternative and having useless ERROR log entries is worse IMHO.

mkouba replied:
> The exception is a `NoStackTraceException` so there is nothing to go by.

Yes, this is something we could report in Vertx maybe? One thing we could do is to also test the connection with `WebSocketConnection.isClosed()`, i.e. `connections.isClosed() && message.contains("WebSocket is closed")` .

> I do agree it's brittle, but I don't see any alternative and having useless ERROR log entries is worse IMHO.

Well, the question is whether it's truly useless (some users might be interested in this) and as I pointed out the other question is how often does it actually happen?

We could also make this configurable, i.e. something like `quarkus.websockets-next.server.log-closed-connection-errors` with default value `false` :shrug:.

geoand replied:
> Well, the question is whether it's truly useless (some users might be interested in this) and as I pointed out the other question is how often does it actually happen?

I easily made it happen in tests. 

I also don't see why developers would need this in their logs, but let's see what @cescoffier thinks as well

geoand replied:
> It's a signal that something went wrong on the server 

Not when the client closed the connection before the server could send out the data

geoand replied:
> even say that it could be seen as a potential security issue if you can easily fill the logs with stacktraces by disconnecting clients

Definitely. 

Also, I am fairly sure we **don't** log these errors (or warnings) in our REST stack when the connection goes away before the response can be written

mkouba replied:
> that we should avoid logging connection errors that can be easily initiated by the client

Like I mentioned above this should only happen infrequently - when the connection is closed during processing of a message or something like that. 

>  I would even say that it could be seen as a potential security issue if you can easily fill the logs with stacktraces by disconnecting clients.

I disagree. In that case, any server error would meet these conditions :thinking:.

But like I mentioned in https://github.com/quarkusio/quarkus/pull/40246#discussion_r1577894970 we should probably log a WARNING instead.

The only problem I can see is that the test relies on an exception message, which is super-brittle, may lead to false positives etc. We should at least combine `(exception instanceof VertxException) && connections.isClosed() && message.contains("WebSocket is closed")`

PS. I'm in the middle of a huge refactoring due to WebSocket client impl and I'm not going to merge any websocket-next PRs until the refactoring is finished :shrug:.

geoand replied:
> I disagree. In that case, any server error would meet these conditions 🤔.

The differentiator is when the user can trigger at will. That is why in Quarkus REST we don't log anything by default.

gsmet replied:
> I disagree. In that case, any server error would meet these conditions

It actually really depends on the conditions.
See for instance: https://github.com/discourse/discourse/security/advisories/GHSA-89h3-g746-xmwq

> The only problem I can see is that the test relies on an exception message, which is super-brittle

That I agree. I would advise that we live with it (with your suggestion included if it works) until something in Vert.x allows us to do better (but we would need to create an issue there)?

geoand committed:
Don't log connection closed exceptions as ERROR in websockets-next

The clients can choose to close at any time, so it does not make
sense to fill up logs with closed connection messages

geoand commented:
PR rebased

mkouba commented on extensions/websockets-next/runtime/src/main/java/io/quarkus/websockets/next/runtime/Endpoints.java:
```diff
@@ -212,14 +211,37 @@ public void handle(Throwable t) {
+    }
+
+    private static boolean isWebSocketIsClosedFailure(Throwable throwable) {
+        if (throwable == null) {
```
Could we add a check that `WebSocketConnectionBase#isClosed()` returns `true`?

I wonder if we should create a Vertx issue to introduce `ConnectionClosedException` or something like that...

geoand replied:
> Could we add a check that WebSocketConnectionBase#isClosed() returns true?

Done

> I wonder if we should create a Vertx issue to introduce ConnectionClosedException or something like that.

@vietj would you be willing to have that in Vertx?

gsmet commented:
If we want this in 3.10, we will need a specific backport as this class has been refactored in 3.11 AFAICS.

geoand commented:
I can do that on Monday

geoand commented:
#40582
